File: 1840-maximum-building-height/1840-maximum-building-height.cpp
class Solution {
public:
    int maxBuilding(int num, vector<vector<int>>& r) {
        r.push_back({1, 0});
        sort(r.begin(), r.end());
        int n = r.size();

        for (int i = 1; i < n; i++)
            r[i][1] = yCap(r[i - 1], r[i]);

        for (int i = n - 2; i >= 0; i--)
            r[i][1] = yCap(r[i + 1], r[i]);

        int res = 0;
        for (int i = 1; i < n; i++)
            res = max(res, yPeak(r[i - 1], r[i]));

        return max(res, r[n - 1][1] + num - r[n - 1][0]);
    }

    int yCap(vector<int>& l, vector<int>& r) {
        int x1 = l[0], y1 = l[1], x2 = r[0], y2 = r[1];
        return min(y2, y1 + abs(x2 - x1));
    }

    int yPeak(vector<int>& l, vector<int>& r) {
        int x1 = l[0], y1 = l[1], x2 = r[0], y2 = r[1];
        return (y1 + y2 + x2 - x1) >> 1;
    }
};

File: 1840-maximum-building-height/1840-maximum-building-height.java
class Solution {
    public int maxBuilding(int num, int[][] restrictions) {
        List<int[]> r = new ArrayList<>(Arrays.asList(restrictions));
        r.add(new int[]{1, 0});
        r.sort((a, b) -> Integer.compare(a[0], b[0]));
        int n = r.size();

        for (int i = 1; i < n; i++)
            r.get(i)[1] = yCap(r.get(i - 1), r.get(i));

        for (int i = n - 2; i >= 0; i--)
            r.get(i)[1] = yCap(r.get(i + 1), r.get(i));

        int res = 0;
        for (int i = 1; i < n; i++)
            res = Math.max(res, yPeak(r.get(i - 1), r.get(i)));

        return Math.max(res, r.get(n - 1)[1] + num - r.get(n - 1)[0]);
    }

    int yCap(int[] l, int[] b) {
        return Math.min(b[1], l[1] + Math.abs(b[0] - l[0]));
    }

    int yPeak(int[] l, int[] b) {
        return (l[1] + b[1] + b[0] - l[0]) >> 1;
    }
}